java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java

```diff
@@ -0,0 +1,476 @@
+        - the symbol/expr to be evaluated
+     */
+
+    int total = 1;
+    BinaryExpressionTree binTree = scootLeft((BinaryExpressionTree) expTree);
+    addSecondaryLocation(flow, binTree, 1, 0);
+
+    while (expTree instanceof BinaryExpressionTree){
+      binTree = scootLeft((BinaryExpressionTree) expTree);
+
+      ExpressionTree left = binTree.leftOperand();
+      ExpressionTree right = binTree.rightOperand();
+
+      if (binTree.kind() != left.kind() && isLogicalOp(left)) {
+        total++;
+        addSecondaryLocation(flow, binTree, 1, 0);
+      }
+      if (binTree.kind() != right.kind() && isLogicalOp(right)) {
+        total++;
+        addSecondaryLocation(flow, binTree, 1, 0);
+      }
+      if (binTree.parent().is(CONDITIONAL_OR) && right.is(CONDITIONAL_AND)) {
+        total++;
+        addSecondaryLocation(flow, binTree, 1, 0);
+      }
+      expTree =  binTree.leftOperand();
+    }
+
+    return total;
+  }
+
+  private static BinaryExpressionTree scootLeft(BinaryExpressionTree binaryExpressionTree) {
+
+    BinaryExpressionTree binTree = binaryExpressionTree;
+
+    if (binTree.is(CONDITIONAL_AND)) {
+      while (binTree.leftOperand().is(CONDITIONAL_AND)) {
+        binTree = (BinaryExpressionTree) binTree.leftOperand();
+      }
+    } else if (binTree.is(CONDITIONAL_OR)) {
+      ExpressionTree left = binTree.leftOperand();
+      while (left.is(CONDITIONAL_OR) && ! isLogicalOp(((BinaryExpressionTree)left).rightOperand())) {
+        binTree = (BinaryExpressionTree) binTree.leftOperand();
+        left = binTree.leftOperand();
+      }
+    }
+    return binTree;
+  }
+
+
+  private static boolean isLogicalOp(ExpressionTree expTree) {
+    return expTree.is(CONDITIONAL_AND) || expTree.is(CONDITIONAL_OR);
+  }
+
+  private int countTryChains(Tree st, List<JavaFileScannerContext.Location> flow, int nestingLevel) {
+    int total = 0;
+
+    if (st.is(Tree.Kind.TRY_STATEMENT)) {
+      TryStatementTree tryStatment = (TryStatementTree) st;
+
+      List<CatchTree> catches = tryStatment.catches();
+      for (CatchTree catchTree : catches) {
+        total++;
+        total += nestingLevel;
+        addSecondaryLocation(flow, catchTree, total, nestingLevel);
+        total += countComplexity(catchTree.block().body(), flow, nestingLevel +1);
+      }
+
+      if (tryStatment.finallyBlock() != null && tryStatment.finallyBlock().body() != null) {
+        total += countComplexity(tryStatment.finallyBlock().body(), flow, nestingLevel);
+      }
+    }
+
+    return total;
+  }
+
+  private static int countBreakAndContinue(Tree st, List<JavaFileScannerContext.Location> flow) {
+    if ( (st.is(BREAK_STATEMENT) && ((BreakStatementTree) st).label() != null)
+            || (st.is(CONTINUE_STATEMENT) && ((ContinueStatementTree) st).label() != null) ) {
+      flow.add(new JavaFileScannerContext.Location("+1", st));
+      return 1;
+
+    }
+
+    return 0;
+  }
+
+  private static Tree getSegmentOfInterest(StatementTree st) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "getSegmentOfInterest" is 16 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)